File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/cassandra/StatementSetterGenerator.java
package ru.tinkoff.kora.database.annotation.processor.cassandra;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.FieldFactory;
import ru.tinkoff.kora.database.annotation.processor.QueryWithParameters;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameter;

import javax.annotation.Nullable;
import javax.lang.model.element.ExecutableElement;
import java.util.List;
import java.util.Objects;

public class StatementSetterGenerator {
    public static void generate(MethodSpec.Builder b, ExecutableElement method, QueryWithParameters sqlWithParameters, List<QueryParameter> parameters, @Nullable QueryParameter batchParam, FieldFactory parameterMappers) {
        if (batchParam != null) {
            b.addCode("var _batch = $T.builder($T.UNLOGGED);\n", CassandraTypes.BATCH_STATEMENT, CassandraTypes.DEFAULT_BATCH_TYPE);
            b.addCode("for (var _param_$L : $L) {$>\n", batchParam.name(), batchParam.name());
        }
        for (int i = 0; i < parameters.size(); i++) {
            var parameter = parameters.get(i);
            if (parameter instanceof QueryParameter.ConnectionParameter) {
                continue;
            }
            var parameterName = parameter.name();
            if (parameter instanceof QueryParameter.BatchParameter batchParameter) {
                parameter = batchParameter.parameter();
                parameterName = "_param_" + parameter.name();
            }
            if (parameter instanceof QueryParameter.SimpleParameter nativeParameter) {
                var isNullable = CommonUtils.isNullable(parameter.variable());
                var sqlParameter = Objects.requireNonNull(sqlWithParameters.find(i));
                if (isNullable) {
                    b.addCode("if ($L == null) {\n", parameter.variable());
                    for (var idx : sqlParameter.sqlIndexes()) {
                        b.addCode("  _stmt.setToNull($L);\n", idx);
                    }
                    b.addCode("} else {$>\n");
                }
                var nativeType = CassandraNativeTypes.findNativeType(ClassName.get(parameter.type()));
                var mapping = CommonUtils.parseMapping(parameter.variable()).getMapping(CassandraTypes.PARAMETER_COLUMN_MAPPER);
                if (nativeType != null && mapping == null) {
                    for (var idx : sqlParameter.sqlIndexes()) {
                        b.addCode(nativeType.bind("_stmt", parameterName, idx)).addCode(";\n");
                    }
                } else if (mapping != null && mapping.mapperClass() != null) {
                    for (var idx : sqlParameter.sqlIndexes()) {
                        var mapper = parameterMappers.get(mapping.mapperClass(), mapping.mapperTags());
                        b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, idx, parameter.variable());
                    }
                } else {
                    for (var idx : sqlParameter.sqlIndexes()) {
                        var mapper = parameterMappers.get(CassandraTypes.PARAMETER_COLUMN_MAPPER, parameter.type(), parameter.variable());
                        b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, idx, parameter.variable());
                    }
                }
                if (isNullable) {
                    b.addCode("$<\n}\n");
                }
            }
            if (parameter instanceof QueryParameter.EntityParameter dtoParameter) {
                for (var field : dtoParameter.entity().columns()) {
                    var isNullable = field.isNullable();
                    var sqlParameter = sqlWithParameters.find(field.queryParameterName(dtoParameter.name()));
                    if (sqlParameter == null || sqlParameter.sqlIndexes().isEmpty()) {
                        continue;
                    }
                    var fieldAccessor = CodeBlock.of("$N.$N()", parameterName, field.accessor()).toString();
                    if (isNullable) {
                        b.addCode("if ($L == null) {\n", fieldAccessor);
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.addCode("  _stmt.setToNull($L);\n", idx);
                        }
                        b.addCode("} else {$>\n");
                    }
                    var nativeType = CassandraNativeTypes.findNativeType(TypeName.get(field.type()));
                    var mapping = CommonUtils.parseMapping(field.element()).getMapping(CassandraTypes.PARAMETER_COLUMN_MAPPER);
                    if (nativeType != null && mapping == null) {
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.addCode(nativeType.bind("_stmt", fieldAccessor, idx)).addCode(";\n");
                        }
                    } else if (mapping != null && mapping.mapperClass() != null) {
                        for (var idx : sqlParameter.sqlIndexes()) {
                            var mapper = parameterMappers.get(mapping.mapperClass(), mapping.mapperTags());
                            b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, idx, fieldAccessor);
                        }
                    } else {
                        for (var idx : sqlParameter.sqlIndexes()) {
                            var mapper = parameterMappers.get(CassandraTypes.PARAMETER_COLUMN_MAPPER, field.type(), field.element());
                            b.addCode("$L.apply(_stmt, $L, $L);\n", mapper, idx, fieldAccessor);
                        }
                    }
                    if (isNullable) {
                        b.addCode("$<\n}\n");
                    }
                }
            }
        }
        if (batchParam != null) {
            b.addStatement("var _builtStatement = _stmt.build()");
            b.addStatement("_batch.addStatement(_builtStatement)");
            b.addCode("_stmt = new $T(_builtStatement);$<\n}\n", ClassName.get("com.datastax.oss.driver.api.core.cql", "BoundStatementBuilder"));
            b.addCode("var _s = _batch.build();\n");
        } else {
            b.addCode("var _s = _stmt.build();\n");
        }
    }

}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/cassandra/StatementSetterGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.cassandra

import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.database.symbol.processor.QueryWithParameters
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameter
import ru.tinkoff.kora.ksp.common.FieldFactory
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.parseMappingData

object StatementSetterGenerator {
    fun generate(
        b: FunSpec.Builder,
        queryWithParameters: QueryWithParameters,
        parameters: List<QueryParameter>,
        batchParam: QueryParameter?,
        parameterMappers: FieldFactory
    ) {
        if (batchParam != null) {
            b.addStatement("val _batch = %T.builder(%T.UNLOGGED)", CassandraTypes.batchStatement, CassandraTypes.defaultBatchType)
            b.beginControlFlow("for (_i in 0 until %N.size)", batchParam.name)
            b.addCode("val _param_%L = %N[_i]\n", batchParam.name, batchParam.name)
        }
        for (i in parameters.indices) {
            var parameter = parameters[i]
            if (parameter is QueryParameter.ConnectionParameter) {
                continue
            }
            var parameterName = parameter.name
            if (parameter is QueryParameter.BatchParameter) {
                parameter = parameter.parameter
                parameterName = "_param_${parameter.name}"
            }
            if (parameter is QueryParameter.SimpleParameter) {
                val sqlParameter = queryWithParameters.find(i)!!
                val isNullable = parameter.type.isMarkedNullable
                if (isNullable) {
                    b.beginControlFlow("if (%N == null)", parameterName)
                    for (idx in sqlParameter.sqlIndexes) {
                        b.addStatement("_stmt.setToNull(%L)", idx)
                    }
                    b.nextControlFlow("else")
                }
                val nativeType = CassandraNativeTypes.findNativeType(parameter.type.toTypeName())
                val mapping = parameter.variable.parseMappingData().getMapping(CassandraTypes.parameterColumnMapper)
                if (nativeType != null && mapping == null) {
                    for (idx in sqlParameter.sqlIndexes) {
                        b.addStatement("%L", nativeType.bind("_stmt", CodeBlock.of("%N", parameterName), CodeBlock.of("%L", idx)));
                    }
                } else if (mapping?.mapper != null) {
                    val mapper = parameterMappers.get(mapping.mapper!!, mapping.tags)
                    for (idx in sqlParameter.sqlIndexes) {
                        b.addStatement("%N.apply(_stmt, %L, %N)", mapper, idx, parameter.variable.name!!.asString())
                    }
                } else {
                    val mapper = parameterMappers.get(CassandraTypes.parameterColumnMapper, parameter.type, parameter.variable)
                    for (idx in sqlParameter.sqlIndexes) {
                        b.addStatement("%N.apply(_stmt, %L, %N)", mapper, idx, parameter.variable.name!!.asString())
                    }
                }
                if (isNullable) {
                    b.endControlFlow();
                }
            }
            if (parameter is QueryParameter.EntityParameter) {
                for (field in parameter.entity.fields) {
                    val parameterNullable = parameter.type.isMarkedNullable
                    val fieldNullable = field.type.isMarkedNullable
                    val sqlParameter = queryWithParameters.find("${parameter.name}.${field.property.simpleName.getShortName()}")
                    if (sqlParameter == null || sqlParameter.sqlIndexes.isEmpty()) {
                        continue
                    }

                    b.addCode("%N", parameterName)
                    if (parameterNullable) b.addCode("?")
                    b.controlFlow(".%N.let {", field.property.simpleName.asString()) {
                        if (parameterNullable || fieldNullable) {
                            b.beginControlFlow("if (it == null)")
                            for (idx in sqlParameter.sqlIndexes) {
                                b.addStatement("_stmt.setToNull(%L)", idx)
                            }
                            b.nextControlFlow("else")
                        }
                        val nativeType = CassandraNativeTypes.findNativeType(field.type.toTypeName());
                        val mapping = field.mapping.getMapping(CassandraTypes.parameterColumnMapper)
                        if (nativeType != null && mapping == null) {
                            for (idx in sqlParameter.sqlIndexes) {
                                b.addStatement("%L", nativeType.bind("_stmt", CodeBlock.of("it"), CodeBlock.of("%L", idx)))
                            }
                        } else if (mapping?.mapper != null) {
                            val mapper = parameterMappers.get(mapping.mapper!!, mapping.tags)
                            for (idx in sqlParameter.sqlIndexes) {
                                b.addStatement("%N.apply(_stmt, %L, it);\n", mapper, idx)
                            }
                        } else {
                            val mapper = parameterMappers.get(CassandraTypes.parameterColumnMapper, field.type, field.property);
                            for (idx in sqlParameter.sqlIndexes) {
                                b.addStatement("%N.apply(_stmt, %L, it);\n", mapper, idx)
                            }
                        }
                        if (parameterNullable || fieldNullable) {
                            b.endControlFlow()
                        }
                    }

                }
            }
        }
        if (batchParam != null) {
            b.addStatement("val _builtStmt = _stmt.build()")
            b.addStatement("_batch.addStatement(_builtStmt)")
            b.addCode("_stmt = %T(_builtStmt)", ClassName("com.datastax.oss.driver.api.core.cql", "BoundStatementBuilder"))
            b.endControlFlow()
            b.addStatement("val _s = _batch.build()")
        } else {
            b.addStatement("val _s = _stmt.build()")
        }
    }
}
